src/main/java/com/mongodb/kafka/connect/util/config/ConfigSoftValidator.java

```diff
@@ -0,0 +1,329 @@
+   *     Map}.
+   * @param topicStrippedProps {@code null} iff {@code topicName} is {@code null}.
+   * @return A {@link Map} where keys are property names (stripped, because {@code
+   *     topicStrippedProps} must contain only stripped property names), and values are pairs of the
+   *     property value and a flag telling whether the property value came from {@code
+   *     topicStrippedProps}, i.e., was overridden, or from {@code globalProps}.
+   */
+  private static Map<String, Entry<String, Boolean>> combineProperties(
+      final Map<String, String> globalProps,
+      @Nullable final String topicName,
+      @Nullable final Map<String, String> topicStrippedProps) {
+    assertTrue((topicName == null) ^ (topicStrippedProps != null));
+    Map<String, Entry<String, Boolean>> combinedStrippedProps = new HashMap<>();
+    globalProps.forEach(
+        (propertyName, propertyValue) ->
+            combinedStrippedProps.put(
+                propertyName, new AbstractMap.SimpleImmutableEntry<>(propertyValue, false)));
+    if (topicStrippedProps != null) {
+      topicStrippedProps.forEach(
+          (propertyName, propertyValue) ->
+              combinedStrippedProps.put(
+                  propertyName, new AbstractMap.SimpleImmutableEntry<>(propertyValue, true)));
+    }
+    return combinedStrippedProps;
+  }
+
+  /**
+   * @param propertyName Either a normal configuration property name, or property name prefixed with
+   *     {@link MongoSinkTopicConfig#TOPIC_OVERRIDE_PREFIX} and a topic name, as specified <a
+   *     href="https://docs.mongodb.com/kafka-connector/master/sink-connector/configuration-properties/topic-override/">here</a>.
+   */
+  private static String strippedPropertyName(final String propertyName) {
+    return propertyName.startsWith(TOPIC_OVERRIDE_PREFIX)
+        ? propertyName.substring(propertyName.indexOf(".", TOPIC_OVERRIDE_PREFIX.length() + 1) + 1)
+        : propertyName;
+  }
+
+  /** @param strippedPropertyName See {@link #strippedPropertyName(String)}. */
+  private static String overriddenPropertyName(
+      final String topicName, final String strippedPropertyName) {
+    return TOPIC_OVERRIDE_PREFIX + topicName + "." + strippedPropertyName;
+  }
+
+  /**
+   * @return {@code ""} if {@code propertyName} is not prefixed with {@link
+   *     MongoSinkTopicConfig#TOPIC_OVERRIDE_PREFIX}, otherwise returns the topic name from {@code
+   *     propertyName}.
+   */
+  private static String topicNameFromPropertyName(final String propertyName) {
+    if (propertyName.startsWith(TOPIC_OVERRIDE_PREFIX)) {
+      int topicNameStartIdx = TOPIC_OVERRIDE_PREFIX.length();
+      return propertyName.substring(
+          topicNameStartIdx, propertyName.indexOf(".", topicNameStartIdx));
+    } else {
+      return "";
+    }
+  }
+
+  private ConfigSoftValidator() {}
+
+  /** A description of a set of obsolete configuration properties. */
+  @Immutable
+  public static final class ObsoletePropertiesSet {
```

Comment:
The `ObsoletePropertiesSet` class is a new piece of code in `ConfigSoftValidator` (the content of `ConfigSoftValidator` was originally in `SinkConfigSoftValidator`).